File: extract_features.py
import pandas as pd
import numpy as np
from imblearn.over_sampling import SMOTE

# read in data from DO
dat = pd.read_stata("data/1_2_3_4A._Done.dta")
dat = dat.set_index('_id')

# collect cat cols
categorical = []
for col in dat.columns:
    if col[0] == "_":
        continue
    if col[1] == "c":
        categorical.append(col)

# convert unordered categorical to dummy
for c in categorical:
    one_hot = pd.get_dummies(dat[c], prefix=c)
    dat = pd.concat([dat, one_hot], axis=1)
    dat = dat.drop(c, axis=1)

# split test and train
test = dat[dat['_cohort'] == 0]
train = dat[dat['_cohort'] == 1]

# estimate mean and SD using training dataset
desc = train.describe()
means = np.array(desc.T['mean'])
stds = np.array(desc.T['std'])

# drop features with SD==0 in training data & all NaN
keep = desc.columns[stds!=0].tolist()
train = train[keep]
test = test[keep]

# sep cols
antenatal = []
antenatal_growth = []
antenatal_intrapartum = []

for col in train.columns:
    if col[0] == "_":
        continue
    if col[0] == "a":
        antenatal.append(col)
        antenatal_growth.append(col)
        antenatal_intrapartum.append(col)
    if col[0] == "g":
        antenatal_growth.append(col)
    if col[0] == "i":
        antenatal_intrapartum.append(col)
        antenatal_growth.append(col)

for name, variable_list in {"antenatal" : antenatal, "antenatal_growth" : antenatal_growth, "antenatal_intrapartum" : antenatal_intrapartum}.items():
    for outcome in ['_hie']:
        print("Working on {} for {}".format(name, outcome))

        # drop features with high covariance; ranked from low to high missingness
        sorted_features = train[variable_list].isnull().sum().sort_values(ascending=True).index.to_list()
        to_drop = set()
        for feature in sorted_features:
            if feature in to_drop:
                continue
            variable_list_tmp = [x for x in variable_list if x not in to_drop and x != feature]
            cor = train[variable_list_tmp].corrwith(train[feature]).abs()
            to_drop.update(cor[cor > 0.95].index.to_list())
        to_keep = [x for x in variable_list if x not in to_drop]
        train_x = train[to_keep]
        test_x = test[to_keep]

        # estimate mean and SD using training dataset
        desc = train_x.describe()
        means = np.array(desc.T['mean'])
        stds = np.array(desc.T['std'])

        # convert to Z score
        train_x = (train_x - means) / stds
        test_x = (test_x - means) / stds

        # drop missing data
        train_x = train_x.join(train[outcome])
        train_x = train_x.dropna(axis='index')
        train_y = train_x.pop(outcome)

        test_x = test_x.join(test[outcome])
        test_x = test_x.dropna(axis='index')
        test_y = test_x.pop(outcome)

        # oversample training dataset using SMOTE
        X_resampled, y_resampled = SMOTE().fit_resample(train_x, train_y)

        # write to csv
        pd.concat([X_resampled, y_resampled], axis=1).to_csv("data/{}{}_train.csv".format(name, outcome), header=True)
        pd.concat([test_x, test_y], axis=1).to_csv("data/{}{}_test.csv".format(name, outcome), header=True)

File: extract_features.R
library("readstata13")
library("dplyr")
library("mltools")
set.seed(124)

# Returns list of highly correlated features to drop
to_drop <- function(tx, features){
    # subset features
    tx <- tx %>% dplyr::select(all_of(features))
    # sort features by missingness
    features <- apply(tx, 2, function(x) sum(is.na(x) | is.infinite(x) | is.nan(x))) %>% sort %>% names
    for (feature in features){
        r <- cor(tx[,feature, with=F], tx[,-feature, with=F], use = "pairwise.complete.obs") %>% abs
        r[r>0.95]
    }
}

# read in data from DO
dat <- read.dta13("data/1_2_3_4A._Done.dta")
rownames(dat) <- dat[['_id']]
dat[['_id']] <- NULL
dat[['_cohort']] <- NULL
dat[['_lapgar']] <- NULL
dat[['_ne']] <- NULL
dat[['_neonataldeath']] <- NULL
dat[['_perinataldeath']] <- NULL
dat[['_resus']] <- NULL
dat[['_stillborn']] <- NULL
dat[['_yearofbirth']] <- NULL

# find categorical variables
cats <- data.frame(id=names(dat)) %>% 
    dplyr::filter(!startsWith(id, "_")) %>% 
    dplyr::filter(substr(id, 2, 2)=="c") %>% 
    dplyr::pull(id)

# factorize categorical variables
dat <- dat %>%
    dplyr::mutate_at(all_of(cats), list(factor))

# convert unordered categorical to dummy
dat <- one_hot(as.data.table(dat))

# split test and train
dat$train <- sample(c(T,F), nrow(dat), replace = T)
train <- dat %>% dplyr::filter(train) %>% dplyr::select(-train)
test <- dat %>% dplyr::filter(!train) %>% dplyr::select(-train)

# split X and Y
train_x <- train %>% dplyr::select(-'_hie')
train_y <- train %>% dplyr::select('_hie')
test_x <- test %>% dplyr::select(-'_hie')
test_y <- test %>% dplyr::select('_hie')

# estimate mean and SD using training dataset
m <- apply(train_x, 2, function(x) mean(x, na.rm=T))
s <- apply(train_x, 2, function(x) sd(x, na.rm=T))

# drop features with SD==0 in training data
train_x <- train_x[,s>0,with=F]
test_x <- test_x[,s>0,with=F]

# Z-score normalisation of predictors
train_x <- (train_x - m) / s
test_x <- (test_x - m) / s

# define feautre sets
antenatal <- data.frame(id=names(train_x)) %>% 
    dplyr::filter(!startsWith(id, "_")) %>% 
    dplyr::filter(substr(id, 1, 1)=="a") %>% 
    dplyr::pull(id)

antenatal_growth <- data.frame(id=names(train_x)) %>% 
    dplyr::filter(!startsWith(id, "_")) %>% 
    dplyr::filter(substr(id, 1, 1)=="a" | substr(id, 1, 1)=="g") %>% 
    dplyr::pull(id)

antenatal_intrapartum <- data.frame(id=names(train_x)) %>% 
    dplyr::filter(!startsWith(id, "_")) %>% 
    dplyr::filter(substr(id, 1, 1)=="a" | substr(id, 1, 1)=="g" | substr(id, 1, 1)=="i") %>% 
    dplyr::pull(id)

# drop features with high covariance
